### ISSUE TITLE
Flutter's Dart SDK version breaks semantic versioning constraints

### ISSUE THREAD

**Issue by tvolkert** [flutter/flutter#14751]
In #14610, we unified the Dart SDK against which the Flutter engine is built with the Dart SDK used to run the `flutter` tool (and used to perform analysis of Flutter app code).

While that change fixed a number of issues, it introduced a problem whereby the Dart SDK version is `2.0.0-edge-<hash>`, which resolves as a newer semver version than `2.0.0-dev-<anything>`.  As the Dart SDK team rolls out breaking SDK changes in the run-up to the Dart 2.0 release, they're cutting newer versions of packages and adding SDK constraints specifying that those newer versions require `2.0.0-dev-xx`.  We're now in a situation where a package may require a newer Dart SDK than Flutter is using, but the semver resolver will think that Flutter's version is newer -- and thus incorrectly cause `pub` to pull down the newer version of the package that's incompatible with Flutter's Dart SDK.

To solve this, we need to make sure the Flutter Dart SDK has a version that reflects its proper placement with respect to tagged dev releases of the Dart SDK.  The solution we're going with is as follows:

```
2.0.0-dev-23.0                     2.0.0-dev-24.0
      ↓                                   ↓
   c123a6    ...     4f8d46    ...     a82b39
                        ↑ 
           2.0.0-dev-23.0-flutter-4f8d46
```

This versioning scheme ensures that the Flutter Dart SDK version:

* Resolve as greater than `2.0.0-dev-23.0` but less than `2.0.0-dev-24.0`
* Allows a casual reader to know approximately what Dart SDK version they're running
* Allows a developer to know what exact Git hash the Dart SDK was built against

Marking this issue as `⚠ TODAY` since any release we cut with the current versioning scheme will be unable to be released to a stable channel.

/cc @Hixie @leafpetersen @a-siva @zanderso @kevmoo @mit-mit @devoncarew @aam

**Comment by zanderso** [comment]
https://dart-review.googlesource.com/c/sdk/+/42026

**Comment by DanTup** [comment]
> `2.0.0-dev-23.0-flutter-4f8d46`

Should some of those dashes be dots? (`2.0.0-dev.23.0.flutter-4f8d46`)?

It was my (possibly bad) understanding of semver that after the first dash things are split on dots then sorted (either numerically or lexically) so using dashes for `dev-x` and `x-flutter` might not do the right thing?

**Comment by DanTup** [comment]
For ex, I think `2.0.0-dev-23.10-flutter-4f8d46` is considered older than `2.0.0-dev-23.9-flutter-4f8d46`?

**Comment by kevmoo** [comment]
I think Dan is right.

Use package:pub_semver to experiment if you want to be 100% sure

**Comment by zanderso** [comment]
In my CL, it is `2.0.0-dev.23.0-flutter-4f8d46`. That should be okay as long as there aren't any dev prerelease patch versions in the next week or so. I'll submit a fix so that it is `2.0.0-dev.23.0.flutter-4f8d46`, but it won't be able to roll into the engine this week.

**Comment by tvolkert** [comment]
With the PR merged, I'm gonna remove the "TODAY" label from this issue since we're no longer blocked from cutting a dev release.  Like @zanderso says, only if there's a dev patch release (unlikely at all, and even more unlikely that any packages will list that patch release in their constraint) will that dev release be affected.
